src/booleans.rs

```diff
@@ -0,0 +1,273 @@
+#[derive(Copy,Clone)]
+pub struct NonminimalBool;
+
+impl LintPass for NonminimalBool {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(NONMINIMAL_BOOL, LOGIC_BUG)
+    }
+}
+
+impl LateLintPass for NonminimalBool {
+    fn check_item(&mut self, cx: &LateContext, item: &Item) {
+        NonminimalBoolVisitor(cx).visit_item(item)
+    }
+}
+
+struct NonminimalBoolVisitor<'a, 'tcx: 'a>(&'a LateContext<'a, 'tcx>);
+
+use quine_mc_cluskey::Bool;
+struct Hir2Qmm<'a, 'tcx: 'a, 'v> {
+    terminals: Vec<&'v Expr>,
+    cx: &'a LateContext<'a, 'tcx>
+}
+
+impl<'a, 'tcx, 'v> Hir2Qmm<'a, 'tcx, 'v> {
+    fn extract(&mut self, op: BinOp_, a: &[&'v Expr], mut v: Vec<Bool>) -> Result<Vec<Bool>, String> {
+        for a in a {
+            if let ExprBinary(binop, ref lhs, ref rhs) = a.node {
+                if binop.node == op {
+                    v = self.extract(op, &[lhs, rhs], v)?;
+                    continue;
+                }
+            }
+            v.push(self.run(a)?);
+        }
+        Ok(v)
+    }
+
+    fn run(&mut self, e: &'v Expr) -> Result<Bool, String> {
+        // prevent folding of `cfg!` macros and the like
+        if !in_macro(self.cx, e.span) {
+            match e.node {
+                ExprUnary(UnNot, ref inner) => return Ok(Bool::Not(box self.run(inner)?)),
+                ExprBinary(binop, ref lhs, ref rhs) => {
+                    match binop.node {
+                        BiOr => return Ok(Bool::Or(self.extract(BiOr, &[lhs, rhs], Vec::new())?)),
+                        BiAnd => return Ok(Bool::And(self.extract(BiAnd, &[lhs, rhs], Vec::new())?)),
+                        _ => {},
+                    }
+                },
+                ExprLit(ref lit) => {
+                    match lit.node {
+                        LitKind::Bool(true) => return Ok(Bool::True),
+                        LitKind::Bool(false) => return Ok(Bool::False),
+                        _ => {},
+                    }
+                },
+                _ => {},
+            }
+        }
+        if let Some((n, _)) = self.terminals
+                                  .iter()
+                                  .enumerate()
+                                  .find(|&(_, expr)| SpanlessEq::new(self.cx).ignore_fn().eq_expr(e, expr)) {
+            #[allow(cast_possible_truncation)]
+            return Ok(Bool::Term(n as u8));
+        }
+        let n = self.terminals.len();
+        self.terminals.push(e);
+        if n < 32 {
+            #[allow(cast_possible_truncation)]
+            Ok(Bool::Term(n as u8))
+        } else {
+            Err("too many literals".to_owned())
+        }
+    }
+}
+
+fn suggest(cx: &LateContext, suggestion: &Bool, terminals: &[&Expr]) -> String {
+    fn recurse(brackets: bool, cx: &LateContext, suggestion: &Bool, terminals: &[&Expr], mut s: String) -> String {
+        use quine_mc_cluskey::Bool::*;
+        match *suggestion {
+            True => {
+                s.extend("true".chars());
```

Comment:
that means it's implemented for anything that iterates over `&str`. so it could be `s.push_str("true")` though.